enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,302 @@
+future work in this area.
+
+## Proposal
+
+The following is a possibly incomplete list of places that we do IP/interface
+selection. Ideally these should all use a single mechanism to do that so they
+are all consistent with each other.
+
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (local DNS) it may not matter if the IP selected is consistent
+with the other cases, but Node IP, configure-ovs, and Keepalived all need to
+match because their functionality depends on it. I'm less familiar with the
+requirements for Etcd, but it seems likely that should match as well. In
+general, it seems best if all IP selection logic comes from one place, whether
+it's strictly required or not.
+
+### Workflow Description
+
+At deployment time the cluster administrator will include a manifest that sets
+KUBELET_NODEIP_HINT appropriately. The nodeip-configuration service (which
+will now be set as a dependency for all other services that need IP/interface
+selection) will use that value to determine the desired IP and interface for
+all services on the node. It will write the results of this selection to a
+well-known location which the other services will consume. This way, we don't
+need to duplicate the selection logic to multiple places. It will happen once
+and be reused as necessary.
+
+#### Example
+
+- resolv-prepender has to run before any other node ip selection can take
+  place. Without resolv.conf populated the nodeip-configuration service cannot
+  pull the runtimecfg image.
+  - Note: This is not relevant for plain UPI deployments, but there are some
+    deployment mechanisms that use the IPI-style resolv.conf management. Since
+    this is not present in all deployments, we will likely need to leave this
+    as it is today. The prepender script uses the runtimecfg node-ip logic,
+    so although it won't be able to consume the output of nodeip-configuration
+    there won't be any duplicate logic either.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
+- When configure-ovs runs, it looks for the interface file written by
+  nodeip-configuration. If found, br-ex will be created on that interface. If
+  not, the existing logic will be used.
+- When keepalived runs it will read the IP from the primary-ip file and the
+  interface from the interface file instead of [re-running the selection
+  logic](https://github.com/openshift/baremetal-runtimecfg/blob/62d6b95649882178ebc183ba4e1eecdb4cad7289/pkg/config/net.go#L11)
+- TODO: Etcd? I believe it has a copy of the IP selection logic from runtimecfg
```

Comment:
as far as i know etcd will generate certificates according to kubelet-ip

Reply:
> as far as i know etcd will generate certificates according to kubelet-ip

I believe this is correct

Reply:
Okay, I'll make a note of that in the next revision.